### DISCUSSION TITLE
[❄️ Snowflake Official] The removal of deprecated grant resources

### DISCUSSION THREAD

**Discussion by sfc-gh-jcieslak** [Snowflake-Labs/terraform-provider-snowflake#2736] [Announcements]
Hey, as you may have already seen in other threads, we announced that with the latest release we concluded [grants redesign](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/ROADMAP.md#redesigning-grants) ([more details](https://github.com/Snowflake-Labs/terraform-provider-snowflake/discussions/1890#discussioncomment-9071073)).
During the grants redesign, we introduced a new set of resources that were made to fully replace the old grant resources. Right now, we don't support deprecated resources and recently we closed all the GitHub issues opened for the deprecated grant resources. As there's an alternative for every deprecated grant resource, we decided to remove the deprecated ones on **June 26, 2024 (Wed)**. You still will be able to use the old grant resources afterward, but only with the older provider versions. Although this will be possible, we always recommend upgrading to the latest provider version as it's the most stable one containing more features, edge cases covered, and bugs resolved. The removal is only about grant resources, every other deprecated resource is planned to be removed in the next major version release (v1.0.0). To support the migration, we recently published a [resource migration guide](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/docs/technical-documentation/resource_migration.md) that describes the process of "no downtime" migration with the help of state manipulation tools.
We are looking forward to hear your thoughts. In case of any errors during the migration, please open a GitHub issue.
Here's the list of resources that we're planning to remove:
- [snowflake_account_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/account_grant)
- [snowflake_database_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/database_grant)
- [snowflake_external_table_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/external_table_grant)
- [snowflake_failover_group_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/failover_group_grant)
- [snowflake_file_format_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/file_format_grant)
- [snowflake_function_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/function_grant)
- [snowflake_integration_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/integration_grant)
- [snowflake_masking_policy_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/masking_policy_grant)
- [snowflake_materialized_view_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/materialized_view_grant)
- [snowflake_pipe_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/pipe_grant)
- [snowflake_procedure_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/procedure_grant)
- [snowflake_resource_monitor_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/resource_monitor_grant)
- [snowflake_role_grants](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/role_grants)
- [snowflake_row_access_policy_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/row_access_policy_grant)
- [snowflake_schema_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/schema_grant)
- [snowflake_sequence_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/sequence_grant)
- [snowflake_stage_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/stage_grant)
- [snowflake_stream_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/stream_grant)
- [snowflake_table_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/table_grant)
- [snowflake_tag_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/tag_grant)
- [snowflake_task_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/task_grant)
- [snowflake_user_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/user_grant)
- [snowflake_view_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/view_grant)
- [snowflake_warehouse_grant](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/warehouse_grant)
- [snowflake_grant_privileges_to_role](https://registry.terraform.io/providers/snowflakedb/snowflake/0.92.0/docs/resources/grant_privileges_to_role)

**Comment by ajbosco** [comment]
This is likely the most hostile change I have ever seen on a project towards users. Why have you made it more difficult for people who actually use your project?

**Comment by tabletcorry** [comment]
Agreed that this is an _extremely_ challenging set of changes to migrate past. Especially frustrating are some of the other changes that appear to be resource name changes for no purpose other than the rename.

All of these changes require a tremendous amount of downstream work which Snowflake appears to be trivializing. State changes are easy to describe in a doc, but scale ver poorly, especially in large deployments.

Very disappointing to see in a previously stable terraform provider.

**Comment by jamiekt** [comment]
It would be useful if you could publish, for each deprecated resource, which resource one is supposed to use instead. Perhaps you have already published that somewhere, if so please could you provide a link to it?

**Reply by sfc-gh-jcieslak** [reply]
Hey, in our registry, we have a section where deprecated resources are listed with their counterparts. As the resources are no longer there, you have to go back to previous versions. Here's the list: https://registry.terraform.io/providers/Snowflake-Labs/snowflake/0.90.0/docs#currently-deprecated-resources. It may be not always correct as the old grant resources had multiple functions in one resource. For example, for granting privileges to shares, we now have `snowflake_grant_privileges_to_share,` and for granting privileges to account role `snowflake_grant_privileges_to_account_role,` but `snowflake_table_grant` ([reference](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/0.90.0/docs/resources/table_grant)) resource had those two functions inside it. Thank you for bringing this up. I'll try to update this post, so It will contain the correct mappings and
we'll try to put this list in several other places so It won't be hidden as it is now. I'll let you know when it will be updated, but for now, try to use the registry link I put above.

**Reply by jamiekt** [reply]
Thank you @sfc-gh-jcieslak. I do think it would be a good idea to update this post with this information, or at least link to https://registry.terraform.io/providers/Snowflake-Labs/snowflake/0.90.0/docs#currently-deprecated-resources as the source of truth.

**Reply by sfc-gh-jcieslak** [reply]
FYI @jamiekt (\+ that's for anyone that is confused about the new grant resources), we added a new section in the grants redesign documentation in the registry that should help with resource mapping: https://registry.terraform.io/providers/snowflakedb/snowflake/latest/docs/guides/grants_redesign_design_decisions#mapping-from-old-grant-resources-to-the-new-ones. 
Sorry that you had to wait for that long @jamiekt , it should be added right with the deprecation/removal announcement.

**Reply by jamiekt** [reply]
That's great, thank you @sfc-gh-jcieslak (Our migration is now complete so too late for us but I'm sure many will benefit from this).

**Comment by denizsurmeli** [comment]
For anyone who are struggling with the transition process, I have some very non-tested, non-verified, pretty messed code that might help with the process: https://github.com/denizsurmeli/terraform-snowflake-migrator

It has helped greatly for the migration in my company, but probably won't work for your case, might give you some ideas on the migration process. The scripts do not have any affiliation with where I work.

**Reply by sthakkar-rdc** [reply]
Hey @denizsurmeli 
How one should use it ?

**Reply by sfc-gh-jcieslak** [reply]
Hey @sthakkar-rdc, you may also want to look at https://github.com/snowflakedb/terraform-provider-snowflake/tree/main/pkg/scripts/migration_script#migration-script if you need any help with grants migration.
